### DISCUSSION TITLE
What are the advantages and disadvantages of different kinds of fingerprints?

### DISCUSSION THREAD

**Discussion by etsi123** [rdkit/rdkit#4177] [General]
I was going through this [post](https://birdlet.github.io/2019/05/25/logs_prediction/) and saw that this person tried multiple different types of fingerprints as features for his machine learning model (e.g., "MACCSKeys", "ErG", "Avalon", "ECFP", "Descriptor", "MACCSKeys+Descriptors","Estate"). Are there some rules of thumb on when one should be applied over the others? What are the advantages/disadvantages of each? Below is the code from the link as it pertains to the ML part, but more info can be found in the link if desired. 

    def fingerprint(mol, fptype="MACCSKeys", radius=2, bits=1024):
        npfp = np.zeros((1,))
        if fptype == "MACCSKeys":
            DataStructs.ConvertToNumpyArray(AllChem.GetMACCSKeysFingerprint(mol), npfp)
        elif fptype == "Avalon":
            DataStructs.ConvertToNumpyArray(GetAvalonFP(mol), npfp)
        elif fptype == "ECFP":
            DataStructs.ConvertToNumpyArray(AllChem.GetMorganFingerprintAsBitVect(mol, radius, bits), npfp)
        elif fptype == "ErG":
            npfp = ReducedGraphs.GetErGFingerprint(mol)
        elif fptype == "Estate":
            npfp = Fingerprinter.FingerprintMol(mol)[0]
        else:
            raise TypeError()
        return npfp

    def descriptors(mol):
        calc=MoleculeDescriptors.MolecularDescriptorCalculator([x[0] for x in Descriptors._descList])
        ds = np.asarray(calc.CalcDescriptors(mol))
        return ds

    mols = [Chem.MolFromSmiles(smi) for smi in smis]
    feature = [np.append(fingerprint(mol), descriptors(mol)) for mol in mols]

    X = np.array(feature)
    y = np.array(logs)

    X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                                    test_size=0.2, 
                                                    random_state=2)
    minfo = mutual_info_regression(X_train, y_train)

**Comment by greglandrum** [comment]
Hi @etsi123 : this is a big topic, but the short answer is that there's no optimal choice. There are multiple publications showing that different fingerprints perform best on different datasets and with different algorithms when doing ML. Here's one we did on the topic a while ago: 
http://pubs.acs.org/doi/abs/10.1021/ci400466r
here's another one that's a bit more recent:
https://doi.org/10.1021/acs.jcim.9b00237
There are many more.

If I had to pick one, and I would really prefer not to have to just pick one!, I think I'd probably use the Morgan fingerprint with radius 2. It tends to do well on many different kinds of problems.
